File: leetcode/python3/permutations-ii.py
#
# @lc app=leetcode id=47 lang=python3
#
# [47] Permutations II
#
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        # ans = []
        # def backtracking(tmp, nums):
        #     if not nums:
        #         if tmp not in ans:
        #             ans.append(tmp)
        #         return
        #     else:
        #         for i in range(len(nums)):
        #           backtracking(tmp + [nums[i]], nums[:i] + nums[i+1:])
        # if nums:
        #     backtracking([], nums)
        # return ans
        ans = []
        def backtracking(tmp, nums):
            if not nums:
                ans.append(tmp)
                return
            else:
                for i in range(len(nums)):
                  if i > 0 and nums[i] == nums[i-1]: continue
                  backtracking(tmp + [nums[i]], nums[:i] + nums[i+1:])
        if nums:
            backtracking([], sorted(nums))
        return ans        

File: leetcode/cpp/permutations-ii.cpp
/*
 * @lc app=leetcode id=47 lang=cpp
 *
 * [47] Permutations II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> ans;
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        helper(nums, 0);
        return ans;
    }

    void helper(vector<int>& nums, int pos) {
        if (pos == nums.size()-1) {
            ans.push_back(nums);
            return;
        }
        for (int i = pos; i < nums.size(); i++) {
            if (i != pos && nums[i] == nums[i-1]) continue;
            swap(nums[i], nums[pos]);
            helper(nums, pos+1);
        }
    }
};
// @lc code=end

